Client/mods/deathmatch/logic/CClientGif.cpp

```diff
@@ -0,0 +1,356 @@
+/*****************************************************************************
+ *
+ *  PROJECT:     Multi Theft Auto
+ *  LICENSE:     See LICENSE in the top level directory
+ *
+ *  Multi Theft Auto is available from https://www.multitheftauto.com/
+ *
+ *****************************************************************************/
+
+#include "StdInc.h"
+#include "CClientGif.h"
+
+#define GIF_ALLOCATE(pointer,size,release) pointer = (uint8_t*)realloc((release) ? 0 : pointer,(release) ? size : 0UL)
+#define BIGE 0
+#define _SWAP(h) ((BIGE) ? ((uint16_t)(h << 8) | (h << 8)) : h)
+
+CClientGif::CClientGif(CClientManager* pManager, ElementID ID, CGifItem* p_GifItem)
+    : ClassInit(this), CClientTexture(pManager, ID, p_GifItem)
+{
+    SetTypeName("gif");
+    m_pManager = pManager;
+    stride = GetRenderItem()->m_uiSizeX * 4;
+    m_pGifDisplay = std::make_unique<CClientGifDisplay>(m_pManager->GetDisplayManager(), this);
+}
+
+CClientGif::~CClientGif(){
+    Unlink();
+}
+
+void CClientGif::Register(std::vector<unsigned char*> frms, std::vector<int> dls) {
+    frames = frms;
+    delays = dls;
+    defaultDelays = dls;
+    m_pGifDisplay->UpdateTexture();
+}
+
+void CClientGif::Next(){
+    if (showing >= GetImageCount()) {
+        showing = 1;
+    }else{
+        showing++;
+    }
+}
+
+void CClientGif::Unlink(){
+    for (unsigned char* frame : frames) {
+        if (frame) {
+            delete[] frame;
```

Comment:
Looks dangerous. Would not it be better to use smart pointers like unique_ptr?

Reply:
how should i use smart pointers for char*? 
like `unique_ptr<unsigend char[]>` ?
if yes then how should i set its size when it is allocating the memory?
i checked it but i didn't see any problems

Reply:
Yes, you can use unique_ptr<unsigend char[]>. But there is a better solution: use std::vector<unsigned char>. It will automatically release all the stuff inside.

Reply:
When working with raw ptrs it's always dangerous.

Reply:
fixed this problem too but it works a little bit slower know
for example it could load a 500x240 gif with 15 frames in 23ms but know it loads it in 78ms